completion/models.py

```diff
@@ -92,17 +92,35 @@ def submit_completion(self, user, course_key, block_key, completion):
                 "block_key must be an instance of `opaque_keys.edx.keys.UsageKey`.  Got {}".format(type(block_key))
             )
         if waffle.waffle().is_enabled(waffle.ENABLE_COMPLETION_TRACKING):
-            obj, is_new = self.get_or_create(  # pylint: disable=unpacking-non-sequence
-                user=user,
-                course_key=course_key,
-                block_type=block_type,
-                block_key=block_key,
-                defaults={'completion': completion},
-            )
+            try:
+                with transaction.atomic():
+                    obj, is_new = self.get_or_create(  # pylint: disable=unpacking-non-sequence
+                        user=user,
+                        course_key=course_key,
+                        block_key=block_key,
+                        defaults={
+                            'completion': completion,
+                            'block_type': block_type,
+                        },
+                    )
+            except IntegrityError:
+                # The completion was created concurrently by another process
+                log.info(
+                    "An IntegrityError was raised when trying to create a BlockCompletion for %s:%s:%s.  "
+                    "Falling back to get().",
+                    user,
+                    course_key,
+                    block_key,
+                )
+                obj = self.get(
+                    user=user,
+                    course_key=course_key,
+                    block_key=block_key,
+                )
             if not is_new and obj.completion != completion:
```

Comment:
If an `IntegrityError` were thrown, wouldn't `is_new` be undefined at this point? I think this will cause an exception.

Reply:
Resolved